raidboss: improvements to ASS/S silkie triggers

wexxlee opened the pull request:
I've attempted to add some triggers for handling Fresh Puff 1 (3 puffs in triangle formation) and Fresh Puff 2 (4 puffs with tethers).  

This is very much a wip - I've tested parts of this with user triggers, but it needs some in-game testing (which I can't do right now because servers are down).  I'm planning to add a trigger for Fresh Puff 3 (2 puffs with effects, 6 "washed" away).  I realistically won't be able to get back to this until Thursday at the earliest though, so I'm throwing this up now in case someone else has a chance to play around with it or comment before then.

For Fresh Puff 2, the predominant strat I've seen is [this](https://i.imgur.com/1Rvv4AL.png).  There is an uptime strat for boss = ice, puffs = cardinals, but it requires coordinated movement between the two players with lightning puffs.  One option is to provide a shorter alert for that scenario without a resolution suggested, but it would be inconsistent with the rest of the trigger.  And the rest of the resolutions are pretty standard (or the only possible resolution).  So I tried to compromise by creating output strings that users could easily modify for this scenario, but.... yeah, definitely open to feedback.

xiashtra commented:
> For Fresh Puff 2, the predominant strat I've seen is [this](https://i.imgur.com/1Rvv4AL.png).

Personally, I've seen another strategy being used more commonly (probably the uptime one you referred to).  This mechanic is going to be tough to call because there are various "correct" ways to solve it, but they rely on everyone following the same plan and fall apart when players follow different plans at the same time.  This might need to be an optional trigger that is disabled by default, like the TEA Wormhole strategy.

wexxlee commented:
> This mechanic is going to be tough to call because there are various "correct" ways to solve it, but they rely on everyone following the same plan and fall apart when players follow different plans at the same time. This might need to be an optional trigger that is disabled by default, like the TEA Wormhole strategy.

I'll give some more thought today to the various options here and try to summarize them for discussion.

xiashtra commented:
If there were a set of basic callouts that indicated the boss action (blue cross or green donut) along with the puff spawn pattern (cardinals or intercardinals) it could be left up to the user to customize the call so it matches their preferred strategy.  E.g., if the default call is something like "boss intercard safe + intercard puffs" and the user changes it to "pull puffs diagonally to corners + stand between yellow and boss" to match the strat diagram you linked for that combination.

wexxlee commented:
> If there were a set of basic callouts that indicated the boss action (blue cross or green donut) along with the puff spawn pattern (cardinals or intercardinals) it could be left up to the user to customize the call so it matches their preferred strategy. E.g., if the default call is something like "boss intercard safe + intercard puffs" and the user changes it to "pull puffs diagonally to corners + stand between yellow and boss" to match the strat diagram you linked for that combination.

I batted it around today, and landed the same place you did.  Not as useful out of the box, but avoids giving incorrect info if a person is using a different strat.  I do think it makes sense to call the boss and then the user's specific tether location/color, rather than just generally "cardinal puffs" or "intercard puffs."  And then have separate output strings for all 8 boss+puffcolor+location combos to let the user fully customize for their strat.  So like:

`
        bait: {
          en: '${boss} - ${dir} ${puff}',
        },
        bossIce: {
          en: 'Boss Blue (intercard safe)',
        },
        bossWind: {
          en: 'Boss Green (middle safe)',
        },
 [...]
        bossIceBlueCardinalPuff: {
          en: 'Blue Puff (bait)',
        },
        bossIceBlueIntercardPuff: {
          en: 'Blue Puff (bait)',
        },
[...]
`

Still not expecting to be able to test this until Thurs, so staying wip for now.

wexxlee commented:
OK, this should be ready for merge consideration.  Added Puff 3, cleaned up some trigger conflicts, and tested in-game.

@xiashtra  I haven't popped into savage yet, so wondering if maybe you have logs or could confirm some ID changes in savage?  The ones we'd need to map (or confirm haven't changed are):

Abilities:
 - 7766: 'Fresh Puff' (from Silkie)
 - 775E: 'Slippery Soap' (from Silkie)
 - 776E: 'Brim Over' (from Eastern Ewer)
 - 776F: 'Rinse' (from Eastern Ewer, repeated; sometimes targets Silken Puff)

Effects (Silken Puffs):
 - CE9: 'Bracing Suds'
 - CEA: 'Chilling Suds'
 - CEB: 'Fizzling Suds'

I'm happy to incorporate these revisions into the savage trigger file with this PR, or defer to you if you prefer to do so.

xiashtra commented:
> OK, this should be ready for merge consideration. Added Puff 3, cleaned up some trigger conflicts, and tested in-game.
> 
> @xiashtra I haven't popped into savage yet, so wondering if maybe you have logs or could confirm some ID changes in savage? The ones we'd need to map (or confirm haven't changed are):
> 
> Abilities:
> 
> * 7766: 'Fresh Puff' (from Silkie)
> * 775E: 'Slippery Soap' (from Silkie)
> * 776E: 'Brim Over' (from Eastern Ewer)
> * 776F: 'Rinse' (from Eastern Ewer, repeated; sometimes targets Silken Puff)
> 
> Effects (Silken Puffs):
> 
> * CE9: 'Bracing Suds'
> * CEA: 'Chilling Suds'
> * CEB: 'Fizzling Suds'
> 
> I'm happy to incorporate these revisions into the savage trigger file with this PR, or defer to you if you prefer to do so.

7766 => 7789
775E => 7781
776E => 7791
776F => 7792

CE9, CEA, CEB no change (status effects seem to be the same between normal and savage in general)

wexxlee commented:
> 7766 => 7789 775E => 7781 776E => 7791 776F => 7792
> 
> CE9, CEA, CEB no change (status effects seem to be the same between normal and savage in general)

Thanks, much appreciated!

wexxlee commented:
Updated with savage triggers/changes.

kshman commented on ui/raidboss/data/06-ew/dungeon/another_sildihn_subterrane-savage.ts:
```diff
@@ -346,6 +473,209 @@ const triggerSet: TriggerSet<Data> = {
+        data.freshPuff2SafeAlert = output[safeOutput]!();
+        return output.bait!({
+          boss: output[silkieStatus]!(),
+          dir: puffDir,
```
missing output[] here?:
dir: output[puffDir]!(),

wexxlee replied:
Yeah, good catch.  I forgot to localize that.

wexxlee commented:
@quisquous Any comments or concerns on this one?

quisquous commented on ui/raidboss/data/06-ew/dungeon/another_sildihn_subterrane-savage.ts:
```diff
@@ -160,19 +227,75 @@ const triggerSet: TriggerSet<Data> = {
+        // See Silken Puff Suds Gain trigger for list of Silken Puff effectIds
+        // By this point, Squeaky Clean Left/Right has changed the N puff and either the SW/SE puff to CE9 (Bracing Suds)
+        // We only care about the unaffected puff's status effect (CEA/CEB) for resolving the mechanic.
+        let stackDir = '';
+        let safeDir = '';
```
These always get assigned before use, so I think(?) they don't need an initialization value.

```suggestion
        let stackDir;
        let safeDir;
```

quisquous commented:
Looks good in general, one tiny comment.  Sorry for the delay in getting back to this.

wexxlee commented:
> Looks good in general, one tiny comment. Sorry for the delay in getting back to this.

No worries!

quisquous commented on ui/raidboss/data/06-ew/dungeon/another_sildihn_subterrane.ts:
```diff
@@ -159,19 +226,75 @@ const triggerSet: TriggerSet<Data> = {
+        // See Silken Puff Suds Gain trigger for list of Silken Puff effectIds
+        // By this point, Squeaky Clean Left/Right has changed the N puff and either the SW/SE puff to CE9 (Bracing Suds)
+        // We only care about the unaffected puff's status effect (CEA/CEB) for resolving the mechanic.
+        let stackDir = '';
+        let safeDir = '';
```
```suggestion
        let stackDir = '';
        let safeDir = '';
```
```suggestion
        let stackDir;
        let safeDir;
```
